cmd/prow-job-dispatcher/main.go

```diff
@@ -0,0 +1,108 @@
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/test-infra/prow/logrusutil"
+)
+
+type options struct {
+	prometheusURL          string
+	prometheusUsername     string
+	prometheusPasswordPath string
+	jobVolumesPath         string
+	updateJobVolumes       bool
+
+	prometheusPassword string
+}
+
+func gatherOptions() options {
+	o := options{}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+
+	fs.StringVar(&o.prometheusURL, "prometheus-url", "https://prometheus-prow-monitoring.apps.ci.l2s4.p1.openshiftapps.com", "The prometheus URL")
+	fs.StringVar(&o.prometheusUsername, "prometheus-username", "", "The Prometheus username.")
+	fs.StringVar(&o.prometheusPasswordPath, "prometheus-password-path", "", "The path to a file containing the Prometheus password")
+	fs.StringVar(&o.jobVolumesPath, "job-volumes-path", filepath.Join(os.TempDir(), "job-volumes.yaml"), "The path to a file containing the job volumes")
+	fs.BoolVar(&o.updateJobVolumes, "update-job-volumes", false, "Update job volumes if set")
+
+	if err := fs.Parse(os.Args[1:]); err != nil {
+		logrus.WithError(err).Fatal("could not parse input")
+	}
+	return o
+}
+
+func (o *options) validate() error {
+	if (o.prometheusUsername == "") != (o.prometheusPasswordPath == "") {
+		return fmt.Errorf("--prometheus-username and --prometheus-password-path must be specified together")
+	}
+	return nil
+}
+
+func (o *options) complete(secrets *sets.String) error {
+	if o.prometheusPasswordPath != "" {
+		bytes, err := ioutil.ReadFile(o.prometheusPasswordPath)
+		if err != nil {
+			return err
+		}
+		o.prometheusPassword = strings.TrimSpace(string(bytes))
+		if o.prometheusPassword == "" {
+			return fmt.Errorf("no content in file: %s", o.prometheusPasswordPath)
+		}
+		secrets.Insert(o.prometheusPassword)
+	}
+	return nil
+}
+
+func main() {
+	secrets := sets.NewString()
+	logrus.SetFormatter(logrusutil.NewCensoringFormatter(logrus.StandardLogger().Formatter, func() sets.String {
+		return secrets
+	}))
+	o := gatherOptions()
+	if err := o.validate(); err != nil {
+		logrus.WithError(err).Fatal("Failed to complete options.")
+	}
+	if err := o.complete(&secrets); err != nil {
+		logrus.WithError(err).Fatal("Failed to complete options.")
+	}
+
+	promClient, err := dispatcher.NewPrometheusClient(o.prometheusURL, o.prometheusUsername, o.prometheusPassword)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to create promethues client.")
+	}
+
+	v1api := v1.NewAPI(promClient)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	var jobVolumes map[string]float64
+	if o.updateJobVolumes {
+		jobVolumes, err = dispatcher.GetJobVolumesFromPrometheus(ctx, v1api)
+		if err != nil {
+			logrus.WithError(err).Fatal("Failed to get job volumes from Prometheus.")
+		}
+		if err = dispatcher.SaveJobVolumes(jobVolumes, o.jobVolumesPath); err != nil {
```

Comment:
How will this file be used? Wouldn't it make sense to make it update the `/sanitize-prow-jobs/_config.yaml` config rather than introducing yet another config file?

Reply:
Save the query result to avoid query each time when running this tool.
We can then control the frequency of fresh the volume map.
And we do not need a working prometheus when running this tool.

Reply:
Yeah but whats the point of storing this intermediate state? Just update the `_config` file of the `sanitize-command` with the results or not?

Reply:
Save the result for the next time to reuse in case that we cannot query prometheus the next time, like prometheus is down, we lose the data of prometheus.
Some old results is better than nothing?

Reply:
Removed.